driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -633,22 +657,47 @@ public void onSet(Connection connection, Message.Response response, long latency
 
                     connection.release();
 
-                    // TODO should we check the response ?
+                    RetryPolicy retryPolicy = retryPolicy();
                     switch (response.type) {
                         case RESULT:
                             if (((Responses.Result)response).kind == Responses.Result.Kind.PREPARED) {
                                 logger.debug("Scheduling retry now that query is prepared");
+                                // do not bother inspecting retry policy, no other decision
+                                // makes sense here
                                 retry(true, null);
                             } else {
-                                logError(connection.address, new DriverException("Got unexpected response to prepare message: " + response));
-                                retry(false, null);
+                                DriverException driverException = new DriverException("Got unexpected response to prepare message: " + response);
+                                RetryPolicy.RetryDecision decision;
+                                if (retryPolicy instanceof ExtendedRetryPolicy)
+                                    decision = ((ExtendedRetryPolicy)retryPolicy).onUnexpectedError(statement, request().consistency(), retriesByPolicy, false);
+                                else
+                                    decision = RetryPolicy.RetryDecision.tryNextHost(null);
+                                if (metricsEnabled()) {
+                                    metrics().getErrorMetrics().getOthers().inc();
+                                    if (decision.getType() == Type.RETRY)
+                                        metrics().getErrorMetrics().getRetriesOnUnexpectedError().inc();
+                                    if (decision.getType() == Type.IGNORE)
+                                        metrics().getErrorMetrics().getIgnoresOnUnexpectedError().inc();
+                                }
+                                processRetryDecision(decision, connection, driverException);
```

Comment:
I wonder if it's worth using the policy here (and in case of error below). Preparing a query is idempotent by nature, so there should be no harm in retrying each time.